network/netcore/src/multiplexing/yamux.rs

```diff
@@ -59,12 +61,36 @@ where
         // Because OnRead mode increases the RTT of window update, bigger buffer size and receive
         // window size perform better.
         config.set_max_buffer_size(MAX_BUFFER_SIZE as usize);
-        config
-            .set_receive_window(RECEIVE_WINDOW)
-            .expect("Invalid receive window size");
-        let socket = Compat::new(socket);
+        config.set_receive_window(RECEIVE_WINDOW);
+        let mut connection = yamux::Connection::new(socket, config, mode);
+        let control = connection.control();
+        let runtime = tokio::runtime::Handle::current();
+        let (mut tx, rx) = mpsc::unbounded();
+        let f = async move {
+            loop {
+                let maybe_substream = connection.next_stream().await.transpose();
+                match maybe_substream {
+                    None => {
+                        return Err(io::Error::new(
+                            io::ErrorKind::BrokenPipe,
+                            "Connection closed",
+                        )) as io::Result<()>;
+                    }
+                    Some(substream) => {
+                        let substream =
+                            substream.map_err(|e| io::Error::new(io::ErrorKind::Other, e));
+                        tx.send(substream)
+                            .await
+                            .map_err(|e| io::Error::new(io::ErrorKind::UnexpectedEof, e))?;
+                    }
+                }
+            }
+        };
+        runtime.spawn(f);
```

Comment:
I would probably suggest changing the API of the StreamMuxer such that you don't have to "hide" the fact that you need to pull for substreams. This also means that you must always use this Muxer inside of a tokio context which dramatically limits where this can be used (e.g if you wanted an in memory connection muxed you still need to use a tokio runtime where as right now you wouldn't have to).

So essentially i'm suggesting to *not* depend on a tokio runtime and the ability to spawn a task on it. It will probably come back to bite us if we did make this assumption. (It also makes it very, awkward when you want to kill the connection but you no longer have access to the task that owns the connection)

Reply:
yes, this was one major decision I wanted to debate and arrive at an agreement on. The lock-in to tokio was my concern as well, and one that I was not too happy about. One concern that I have is tying the SubstreamMuxer trait too closely to the Yamux implementation. Is there value in having this as a trait at all?

> It also makes it very, awkward when you want to kill the connection but you no longer have access to the task that owns the connection

Killing the connection is actually not a problem, and can be done using the `Control` object which is `Clone`able.